### ISSUE TITLE
Feature request: Clone all Layers with custom drawings to the next existing page

### ISSUE THREAD

**Issue by flo2k** [xournalpp/xournalpp#1911]
I do use xournal++ to annotate my PDF slides which I created with the latex beamer package. I heavily use the \pause and \visibility features to uncover the individual items. The resulting PDF with the slides contains one page per uncovered item. When I draw (annotate) something that belongs to item 1 and I uncover item 2, which means that I go the next page, then my custom drawings (annotations) are away. I would like to have a mechanism that allows me to reuse my previous drawings (annotations) on the next page easily.

Two buttons with: copy drawings (annotations) (with all layers - except of the background layer) to the next or from the previous slide would be perfect. 
I do also need shortcuts for that to be able to use it from my wacom tablet with the express keys.

I would really appreciate such a feature. Maybe this is already possible, but I don't know how.

**Comment by LittleHuba** [comment]
Duplicate of #1647

**Comment by Technius** [comment]
Not a duplicate--there is currently no way to duplicate or copy/paste layers.

**Comment by flo2k** [comment]
Thank you for the hint to https://github.com/xournalpp/xournalpp/issues/1647. I think the problem is the same: we would like to have the existing drawings (annotations) within one slide (which can consist of several pages) visible.

My first approach was to think that I can do this manually by copying the existing annotations from the previous page to the current, or from the current to the next page as I described above.

But if xournal++ automatically detects that the slide consists of several pages than this is definitely also a good solution for me, if not the better one.

**Comment by rolandlo** [comment]
What you currently can do in order to clone all handwritten layers to the next page (which is assumed to be without annotations): 

1. copy the whole page including background (in the sidebar), 
2. replace the background of the copied page with the next page background (from the menu journal->background)
3. delete the next page without annotations (in the sidebar)

**Comment by flo2k** [comment]
@rolandlo, you are right. This is a solution if you need this only once. However, if you are in a lecture where you need this several times, then this is not a good solution.

**Comment by michaelcadilhac** [comment]
I've spent a couple hours doing something along these lines, precisely for the same reasons.  This needs testing and proper icons, but this fits my needs so far.  The features are:
- Copy current layer to next page
- Copy all (equivalent to some "select all + copy")
- Paste in place; this allows things that live at one position on a slide to be pasted at the same position on another slide.

See https://github.com/michaelcadilhac/xournalpp/tree/incremental-layers

**Comment by rolandlo** [comment]
@michaelcadilhac Great job! This will certainly be very useful for all those Xournal++ users creating slides. Some remarks:

- Undo/redo operations do not quite work yet as expected.
- Lots of things have changed with respect to selections in the snapping enhancements PR #2011, which got merged today. You may want to rebase your work on the latest master therefore before continuing.
- I understand your question about the elements removed from the layer (which does not exist on a page in your case), but I haven't figured out the best solution so far. I will think about it.
- Would you please provide a PR to the official xournalpp master branch? It would simplify communication.
- Copying more than one layer at one time will be more complicated. If you create slides one by one the next page will usually only have the background layer. The layers would first have to be created on the new page. But it is also possible that a user has already created some layers on the next page and decides to clone the layers from the previous page only then. It's not quite clear what should be done in that case. Maybe users have different preferences.
- It is possible to use plugins to get some user specific behavior and shortcuts. Maybe cloning all (!) handwritten layers to the next page could also be made via a plugin that combines the 3 actions I listed in my last post. The plugin interface will need to be extended for that, but it doesn't look too hard. @flo2k What do you think about that?

**Comment by rolandlo** [comment]
I have extended the plugin interface, making the following Lua code copy the handwritten layers of a page with pdf background to the next page, assumed to have pdf background and no annotations yet:

```lua

function clone()
  app.sidebarAction("COPY");
  app.changeBackgroundPdfPageNr(1, true);         
  app.uiAction({["action"]="ACTION_GOTO_NEXT"});
  app.uiAction({["action"]="ACTION_DELETE_PAGE"});
  app.uiAction({["action"]="ACTION_GOTO_BACK"});
end
```

In short: the current page gets copies, receives the background from the next page and the obsolete next page is deleted. It can be called with a shortcut from the keyboard.

The `app.sidebarAction` and `app.changeBackgroundPdfPageNr` functions are new. The second argument is a boolean which indicates if the first argument (page number) is relative (or not). 

This approach has some shortcomings. Most notably it works only as expected if all pages have pdf background and if the page following the one from which you clone has no annotations yet. 
Anyway it may already be useful for annotating LaTeX beamer slides. Any opinions?

**Comment by michaelcadilhac** [comment]
Thanks @rolandlo, I'll be working on a PR based on current master, if you think this is a valid starting point.  Note that because of the CopyAll incorrect logic, this introduces crashes, so maybe we'd want to have a way (any way) to fix this before I make a PR.

As for the Lua snippet: I don't entirely understand what it is doing, but just to be sure, we have two use cases with beamer slides:
1. Duplicate the current page, including layers and background, by inserting a new page.
2. Copy all the handwritten layers to the next page, without inserting a new page.

That's because we combine handwritten animations (1), where the annotation changes but not the slide, and slide animations (2), where the slide changes but not the annotation.  Is your clone function a third use case that I missed?

**Comment by rolandlo** [comment]
It's case (2). See this little demo:

![cloneLayersDemo](https://user-images.githubusercontent.com/40485433/93860151-eee21080-fcbe-11ea-984f-82d7f13fdff1.gif)

The first time I used the menu to activate the plugin, the other times I used the keyboard shortcut. Scrolling to the next page does not yet work properly, that's why I displayed it all in one row.

Sure, we can first fix the bug, before you make a PR. For Xournal++ it is not unusual, however, to have work in progress PR's that are not supposed to be ready for merge (or close to it) yet.

**Comment by rolandlo** [comment]
Now everything works as expected with my plugin (using an extended plugin API). Check it out in PR #2406.
